base/util/factory.h

```diff
@@ -0,0 +1,28 @@
+#pragma once
+
+#include <map>
+#include <memory>
+
+#include "base/context/context.h"
+
+namespace vertexai {
+
+template <typename Product>
+class FactoryRegistrar final {
+ public:
+  typedef std::function<std::unique_ptr<Product>(const context::Context& ctx)> Factory;
+
+  static FactoryRegistrar<Product>* Instance() {
+    static FactoryRegistrar<Product> instance;
+    return &instance;
+  }
+
+  void Register(Factory factory, int priority = 0) { factories_.insert(std::make_pair(priority, factory)); }
+
+  std::multimap<int, Factory> Factories() { return factories_; }
+
+ private:
+  std::multimap<int, Factory> factories_;
```

Comment:
Just a minor thing, fixable later -- it'd be nice to have an enum for the priorities.  (Maybe `Product::Priority`?)

Reply:
This is addressed in an upcoming Metal HAL PR.